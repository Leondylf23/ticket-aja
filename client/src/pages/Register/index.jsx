import { connect, useDispatch } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { FormattedMessage, useIntl } from 'react-intl';
import { useNavigate } from 'react-router-dom';
import PropTypes from 'prop-types';

import { useEffect, useState } from 'react';
import { sendRegisterData, showPopup } from '@containers/App/actions';
import { encryptDataAES } from '@utils/allUtils';
import { selectLogin } from '@containers/Client/selectors';

import classes from './style.module.scss';

const Register = ({ isLogin }) => {
    const intl = useIntl();
    const navigate = useNavigate();
    const dispatch = useDispatch();

    const [fullname, setFullname] = useState('');
    const [email, setEmail] = useState('');
    const [password, setPassword] = useState('');
    const [dob, setDob] = useState('');
    const [role, setRole] = useState('');

    const sendRegister = (e) => {
        e.preventDefault();

        if (fullname.length < 3 || fullname.length > 255) {
            dispatch(showPopup(intl.formatMessage({ id: 'register_title' }), intl.formatMessage({ id: 'register_fullname_validation' })));
            return;

        } else if (dob === '') {
            dispatch(showPopup(intl.formatMessage({ id: 'register_title' }), intl.formatMessage({ id: 'register_dob_validation' })));
            return;

        } else if ((new Date().getTime - new Date(dob).getTime()) < (15 * 365 * 24 * 3600000)) {
            dispatch(showPopup(intl.formatMessage({ id: 'register_title' }), intl.formatMessage({ id: 'register_dob_age_validation' })));
            return;

        } else if (!(/^[^\s@]+@[^\s@]+\.[^\s@]+$/).test(email)) {
            dispatch(showPopup(intl.formatMessage({ id: 'register_title' }), intl.formatMessage({ id: 'register_email_validation' })));
            return;

        } else if (password.length < 6 || password.length > 20) {
            dispatch(showPopup(intl.formatMessage({ id: 'register_title' }), intl.formatMessage({ id: 'register_password_validation' })));
            return;

        } else if (role === '') {
            dispatch(showPopup(intl.formatMessage({ id: 'register_title' }), intl.formatMessage({ id: 'register_role_valiation' })));
            return;

        }

        const formData = {
            fullname: encryptDataAES(fullname),
            email: encryptDataAES(email),
            password: encryptDataAES(password),
            dob: encryptDataAES(dob),
            role: encryptDataAES(role)
        };

        dispatch(sendRegisterData(formData, () => {
            navigate('/login');
        }, (err) => {
            if (err?.response?.status === 422) {
                dispatch(showPopup(intl.formatMessage({ id: 'register_title' }), intl.formatMessage({ id: 'register_email_has_taken' })));
            } else {
                dispatch(showPopup());
            }
        }));
    };

    useEffect(() => {
        if (isLogin) navigate('/');
    }, [isLogin]);

    return (
        <div className={classes.fullContainer}>
            <div className={classes.innerContainer}>
                <h1 className={classes.title}><FormattedMessage id='register_title' /></h1>
                <form className={classes.formContainer} onSubmit={sendRegister}>
                    <label htmlFor='fullname' className={classes.label}><FormattedMessage id='register_fullname' /></label>
                    <input type='text' id='fullname' className={classes.input} placeholder={intl.formatMessage({ id: 'register_fullname_placeholder' })} onChange={(e) => setFullname(e.target.value)} />
                    <label htmlFor='dob' className={classes.label}><FormattedMessage id='register_dob' /></label>
                    <input type='date' id='dob' className={classes.input} onChange={(e) => setDob(e.target.value)} />
                    <label htmlFor='email' className={classes.label}><FormattedMessage id='register_email' /></label>
                    <input type='email' id='email' className={classes.input} placeholder={intl.formatMessage({ id: 'register_email_placeholder' })} onChange={(e) => setEmail(e.target.value)} />
                    <label htmlFor='password' className={classes.label}><FormattedMessage id='register_password' /></label>
                    <input type='password' id='password' className={classes.input} placeholder={intl.formatMessage({ id: 'register_password_placeholder' })} onChange={(e) => setPassword(e.target.value)} />
                    <label htmlFor='role' className={classes.label}><FormattedMessage id='register_role' /></label>
                    <select name='role' id='role' className={classes.input} onChange={(e) => setRole(e.target.value)}>
                        <option value=''><FormattedMessage id='register_role_placeholder' /></option>
                        <option value='customer'><FormattedMessage id='register_role_customer' /></option>
                        <option value='business'><FormattedMessage id='register_role_business' /></option>
                    </select>
                    <button type='submit' className={classes.button}><FormattedMessage id='register_button' /></button>
                    <h3 className={classes.footer}>
                        <FormattedMessage id='register_login_footer' />
                        <a onClick={() => navigate('/login')} className={classes.footerLink}><FormattedMessage id='register_login_footer_link' /></a>
                        .
                    </h3>
                </form>
            </div>
        </div>
    );
};

Register.propTypes = {
    isLogin: PropTypes.bool
}

const mapStateToProps = createStructuredSelector({
    isLogin: selectLogin
});

export default connect(mapStateToProps)(Register);