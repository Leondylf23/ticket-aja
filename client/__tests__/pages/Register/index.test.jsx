import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import RouterDom from 'react-router-dom';

import Register from '@pages/Register';
import store from '@store';
import Language from '@containers/Language';


jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux')
}));

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: jest.fn()
}));

jest.mock('reselect', () => ({
    ...jest.requireActual('reselect')
}));

const ParentComponent = (children) => (
    <Provider store={store}>
        <Language>
            {children}
        </Language>
    </Provider>
);

// const navigate = jest.fn().mockName('navigate');

describe('Register Page', () => {
    // beforeEach(() => {
    //     jest.spyOn(RouterDom, 'useNavigate').mockImplementation(() => navigate);
    // });

    test('Rendered', () => {
        const { getByTestId } = render(ParentComponent(
            <Register />
        ));
        const registerPage = getByTestId('register-page');
        expect(registerPage).toBeInTheDocument();
    });

    test('Should match with snapshot', () => {
        const registerPage = render(ParentComponent(
            <Register />
        ));
        expect(registerPage).toMatchSnapshot();
      });
});
