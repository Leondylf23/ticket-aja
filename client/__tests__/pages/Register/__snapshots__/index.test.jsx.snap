// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Page Should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="fullContainer"
        data-testid="login-page"
      >
        <div
          class="innerContainer"
        >
          <h1
            class="title"
          >
            Masuk
          </h1>
          <form
            class="formContainer"
          >
            <label
              class="label"
              for="email"
            >
              Email
            </label>
            <input
              class="input"
              id="email"
              placeholder="contoh_email@mail.com"
              type="email"
            />
            <label
              class="label"
              for="password"
            >
              Kata Sandi
            </label>
            <input
              class="input"
              id="password"
              placeholder="p4sSW0rdR4haS1a"
              type="password"
            />
            <div
              class="forgotPasswordContainer"
            >
              <a
                class="text"
              >
                Lupa Kata Sandi?
              </a>
            </div>
            <button
              class="button"
              type="submit"
            >
              Masuk
            </button>
            <h3
              class="footer"
            >
              Belum punya akun? Daftar di 
              <a
                class="footerLink"
              >
                sini
              </a>
              .
            </h3>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="fullContainer"
      data-testid="login-page"
    >
      <div
        class="innerContainer"
      >
        <h1
          class="title"
        >
          Masuk
        </h1>
        <form
          class="formContainer"
        >
          <label
            class="label"
            for="email"
          >
            Email
          </label>
          <input
            class="input"
            id="email"
            placeholder="contoh_email@mail.com"
            type="email"
          />
          <label
            class="label"
            for="password"
          >
            Kata Sandi
          </label>
          <input
            class="input"
            id="password"
            placeholder="p4sSW0rdR4haS1a"
            type="password"
          />
          <div
            class="forgotPasswordContainer"
          >
            <a
              class="text"
            >
              Lupa Kata Sandi?
            </a>
          </div>
          <button
            class="button"
            type="submit"
          >
            Masuk
          </button>
          <h3
            class="footer"
          >
            Belum punya akun? Daftar di 
            <a
              class="footerLink"
            >
              sini
            </a>
            .
          </h3>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
